src/storage/engine/mod.rs

```diff
@@ -33,7 +33,17 @@ pub const TEMP_DIR: &'static str = "";
 const SEEK_BOUND: usize = 30;
 const DEFAULT_TIMEOUT_SECS: u64 = 5;
 
-pub type Callback<T> = Box<FnBox(Result<T>) + Send>;
+pub type Callback<T> = Box<FnBox((CbContext, Result<T>)) + Send>;
+
+pub struct CbContext {
```

Comment:
I'm not saying reuse the `Context`. You can check the `Result` to get the actual term.